Proposal for a reference string to number conversion facility in stdlib (#743)
fortran-lang/stdlib: doc/specs/stdlib_str2num.md

---
title: str2num
---

# The `stdlib_str2num` module

This module proposes a function-style interface for string-to-number conversion. It also profits from Fortran's interfaces to implement precision-dependant algorithms to maximize runtime efficiency.

[TOC]

## `to_num` - conversion of strings to numbers

### Status

Experimental

### Description

Convert a string or an array of strings to numerical types.

### Syntax

`number = [[stdlib_str2num(module):to_num(interface)]](string, mold)`

### Arguments

`string`: argument has `intent(in)` and is of type `character(*)`.

`mold`: argument has `intent(in)` and is of numerical type (that is of `integer` or of `real`). **Note**: The type of the `mold` argument defines the type of the result.

### Return value

Return a scalar of numerical type (i.e., `integer`, or `real`).

### Example

```fortran
program example_string_to_number
  use stdlib_kinds, only: dp
  use stdlib_str2num, only: to_num
  implicit none
  character(:), allocatable :: txt
  real(dp) :: x

  txt = ' 8.8541878128e−12 '
  x = to_num( txt , x )

> end program example_string_to_number

Author: Done in the last two commits: https://github.com/fortran-lang/stdlib/pull/743/commits/fbdf87033ccdd1825e9035344a9d8637a92fff73 and https://github.com/fortran-lang/stdlib/pull/743/commits/cfb95daf00807d11970e2e5fd334fc56459197fc

---
title: str2num
---

# The `stdlib_str2num` module

This module proposes a function-style interface for string-to-number conversion. It also profits from Fortran's interfaces to implement precision-dependant algorithms to maximize runtime efficiency.

[TOC]

## `to_num` - conversion of strings to numbers

### Status

Experimental

### Description

Convert a string or an array of strings to numerical types.

### Syntax

`number = [[stdlib_str2num(module):to_num(interface)]](string, mold)`

### Arguments

`string`: argument has `intent(in)` and is of type `character(*)`.

> `mold`: argument has `intent(in)` and is of numerical type. currently: `integer`, `real32`, `real64` or `real128`. **Note**: the mold argument is included to help compilers chose the correct implementation at compile-time. Currently, compilers are not able to disambiguate functions with respect to the left-hand-side of an assignment.

Reviewer: All interpretation of expressions in Fortran and most other programming languages is context-free.  This is not a temporary inability of compilers to perform an analysis, and there are far more possible contexts than being the entire right-hand side of an expression.

Reviewer 2: Actually, because all types of `integer` and `real` can be now supported (through fypp preprocessing) the following should be sufficient (similarly as in other specs):
```suggestion
`mold`: argument has `intent(in)` and is of numerical type (that is of `integer` or of `real32`). **Note**: the mold argument is included to help compilers chose the correct implementation at compile-time. Currently, compilers are not able to disambiguate functions with respect to the left-hand-side of an assignment.
```

Author: Agreed, I felt some comment was needed in order to justify the requirement of a mold argument. What would you suggest as a modification to this comment?

Reviewer 2: Additionally, with the latest changes, kinds are defined in `stdlib_kinds` and not in `iso_fortran_env`, where `stdlib_kinds` defines its own `sp`, `dp` and `qp`.

Reviewer: "The type of the `mold` argument defines the type of the result."

### Return value

> Return a scalar of numerical type (`integer`, `real32` or `real64`).

Reviewer 2: ```suggestion
Return a scalar of numerical type (i.e., `integer`, or `real`).
```

### Example

```fortran
program example_string_to_number
  use stdlib_kinds, only: dp

>   use stdlib_str2num

Reviewer 2: ```suggestion
  use stdlib_str2num, only: to_num
```

  implicit none
  character(:), allocatable :: txt
  real(dp) :: x

  txt = ' 8.8541878128e−12 '
  x = to_num( txt , x )

> end program example_random_seed

Reviewer 2: ```suggestion
end program example_string_to_number
```

```

## `to_num_p` - conversion of a stream of values in a string to numbers

### Status

Experimental

### Description

Convert a stream of values in a string to an array of values.

### Syntax

> `number = [[stdlib_str2num(module):to_num_p(interface)]](string, mold)`

Reviewer 2: Suggestion:

transform it to a subrouine and hide the pointer and the loop inside the subroutine:
```fortran
call to_num_p(string, array)
...
subroutine to_num_p(string, array)
character(*), intent(in), target :: string
real(dp), intent(out) :: array(:)
....
cptr => string

do i = 1, size(array)
...
enddo

end subroutine
```

Author: I've been giving some thought to your idea, on one side I think it does facilitate things, but at the same time I feel that it would introduce some particular cases to handle (multiple dimension arrays, stopping on some condition other than the end of the string, etc) providing the low level interface gives the user the freedom to adapt on top of it. What do you think?

Reviewer 2: Yes, it makes sense to provide the low-level interface only. Keep it as is for now

### Arguments

> `string`: argument has `intent(in)` and is of type `character(*), pointer`.

Reviewer 2: ```suggestion
`string`: argument has `intent(in)` and is of type `character(:), pointer`.
```

> `mold`: argument has `intent(in)` and is of numerical type. currently: `integer`, `real32`, `real64` or `real128`. **Note**: the mold argument is included to help compilers chose the correct implementation at compile-time. Currently, compilers are not able to disambiguate functions with respect to the left-hand-side of an assignment.

Reviewer 2: What about  `integer(int8)`, `integer(int64)`?

> `mold`: argument has `intent(in)` and is of numerical type. currently: `integer`, `real32`, `real64` or `real128`. **Note**: the mold argument is included to help compilers chose the correct implementation at compile-time. Currently, compilers are not able to disambiguate functions with respect to the left-hand-side of an assignment.

Reviewer 2: ```suggestion
`mold`: argument has `intent(in)` and is of numerical type. currently: `int8`, `int16`,`int32`, `int64`, `real32`, `real64` or `real128`. **Note**: the mold argument is included to help compilers chose the correct implementation at compile-time. Currently, compilers are not able to disambiguate functions with respect to the left-hand-side of an assignment.
```

> `mold`: argument has `intent(in)` and is of numerical type. currently: `int8`, `int16`,`int32`, `int64`, `real32`, `real64` or `real128`. **Note**: the mold argument is included to help compilers chose the correct implementation at compile-time. Currently, compilers are not able to disambiguate functions with respect to the left-hand-side of an assignment.

Reviewer 2: ```suggestion
`mold`: argument has `intent(in)` and is of numerical type (currently of `integer` or `real`). **Note**: the mold argument is included to help compilers chose the correct implementation at compile-time. Currently, compilers are not able to disambiguate functions with respect to the left-hand-side of an assignment.
```

### Return value

> Return a scalar of numerical type (`integer`, `real32` or `real64`).

Reviewer 2: ```suggestion
Return a scalar of numerical type (`int8`, int16`, int32`, int64`, `real32`, `real64` or `real128`).
```
I wonder if we should use e.g., `sp` instead of `real32`.

Author: Wonder the same thing. For the doc I think the most explicit would be the best.

Reviewer 2: See comment below: Shortly, the following could be written (similarly to other specs):
```suggestion
Return a scalar of numerical type (i.e., `integer` or `real`).
```

this will be not fypp/compiler-dependent

### Example

```fortran

> program example_str2num

Reviewer 2: ```suggestion
program example_to_num_p
```

    use stdlib_kinds, only: dp

>     use stdlib_str2num

Reviewer 2: ```suggestion
    use stdlib_str2num, only: to_num_p
```

    character(:), allocatable, target :: chain
    character(len=:), pointer :: cptr

>     real(dp), allocatable :: r(:)
>     integer :: i 
> 
>     chain = " 1.234   1.E1 1e0     0.1234E0  12.21e+001 -34.5E1"
>     allocate( r(6) )

Reviewer 2: ```suggestion
    real(dp) :: r(6)
    integer :: i 

    chain = " 1.234   1.E1 1e0     0.1234E0  12.21e+001 -34.5E1"
```

    cptr => chain

>     do i =1, 6

Reviewer 2: ```suggestion
    do i = 1, 6
```

>         r(i) = to_num_p( cptr , r(i) ) !> the cptr pointer is shifted within the function

Reviewer 2: Why do you use a `c_ptr`, and just provide the `character(*)` as an argument (and potentially use a pointer inside the function? This would facilitate its use by the user.

    end do

> end program

Reviewer 2: ```suggestion
end program example_to_num_p
```

```

## Note

> The accuracy of the conversion is implementation dependent; it is recommend that implementers guarantee precision down to the last 3 bits.

Reviewer 2: ```suggestion
The accuracy of the conversion is implementation dependent; it is recommended that implementers guarantee precision down to the last 3 bits.
```

How can the implementers guarantee that?

Author: I wrote this after a discussion with @ivan-pi as a kind of recommendation, but it is difficult to provide a hard rule on how to guarantee it. This is more or less the current upper-bound of the error for qp. Maybe someone can come-up with a better implementation down the road to improve this :)

**The current implementation has been tested to provide for** :

> `real32`  : exact match
> 
> `real64`  : precision up-to epsilon(0.0_real64)
> 
> `real128` : precision around 200*epsilon(0.0_real128)

Reviewer 2: ```suggestion
`sp`  : exact match

`dp`  : precision up-to epsilon(0.0_dp)

`qp` : precision around 200*epsilon(0.0_qp)
```
